File: 040/combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function combinationSum2(candidates, target) {
    candidates.sort((a, b) => a - b);
    let res = [];
    let out = [];

    function combinationSumHelper(candidates, target, start, out, res) {
        if (target < 0) return;
        else if (target == 0) res.push(out.map(i => i));
        else {
            for (let i = start; i < candidates.length; i++) {
                if (i > start && candidates[i] == candidates[i-1]) continue;
                out.push(candidates[i]);
                combinationSumHelper(candidates, target - candidates[i], i + 1, out, res);
                out.pop();
            }
        }
    };

    combinationSumHelper(candidates, target, 0, out, res);
    return res;
}

let candidates = [2, 5, 2, 1, 2], target = 5;
console.log(combinationSum2(candidates, target));

File: 040/combination-sum-ii.go
package main

import (
	"fmt"
	"sort"
)

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)

	var res [][]int
	var solution []int
	combinationSum2Helper(candidates, target, solution, &res)

	return res
}

func combinationSum2Helper(candidates []int, target int, solution []int, result *[][]int) {
	if target == 0 {
		*result = append(*result, solution)
	}

	if len(candidates) == 0 || target < candidates[0] {
		return
	}

	solution = solution[:len(solution):len(solution)]

	combinationSum2Helper(candidates[1:], target-candidates[0], append(solution, candidates[0]), result)
	combinationSum2Helper(next(candidates), target, solution, result)
}

func next(candidates []int) []int {
	i := 0
	for i+1 < len(candidates) && candidates[i] == candidates[i+1] {
		i++
	}
	return candidates[i+1:]
}

func main() {
	candidates, target := []int{10, 1, 2, 7, 6, 1, 5}, 8
	fmt.Println(combinationSum2(candidates, target))
}
